### PR TITLE
Lots of fixes

### PR DESCRIPTION
**Pull request by PatrickRudolph**
* Make it useful on recent Intel platforms with all of it's blobs.
* Add support for inserting ELF payloads. Will be extended with EDK2 and Linux bzImage support next.
* Pretty print image content, error messages and cmdline arguments

### CODE REVIEW

File: cmds/cbfs/cbfs.go [hunk 1]
```diff
@@ -1,37 +1,218 @@
...
+type CommandlineArgs struct {
+	Debug           bool
+	Image           string
+	Command         string
+	CustomArguments map[string]*CustomCommandlineArg
+}
 
-func main() {
-	flag.Parse()
+type CustomCommandlineArg struct {
```
Comment (rminnich) [inline cmds/cbfs/cbfs.go]: run go vet and golint and do what it says :-) it will complain about this.

File: cmds/cbfs/cbfs.go [hunk 2]
```diff
@@ -1,37 +1,218 @@
...
+				return fmt.Errorf("CBFS file name not set")
+			}
+
+			hdrs, data, err := cbfs.ELFToPayload(args.CustomArguments["filename"].Value, cbfs.LZ4)
+			if err != nil {
+				return fmt.Errorf("Failed to convert ELF to payload: %v", err)
+			}
+
+			buf := new(bytes.Buffer)
+			err = binary.Write(buf, binary.BigEndian, hdrs)
```
Comment (rminnich) [inline cmds/cbfs/cbfs.go]: Always check all errors.

File: cmds/cbfs/cbfs.go [hunk 3]
```diff
@@ -1,37 +1,218 @@
...
 
-	if *debug {
-		cbfs.Debug = log.Printf
+			rw, err := cbfs.NewPayloadRecord(&f)
+			if err != nil {
+				return fmt.Errorf("failed to create payload record. Error: %v", err)
+			}
+			rw.Read(bytes.NewReader(buf.Bytes()))
+
+			err = i.AddFile(rw)
```
Comment (rminnich) [inline cmds/cbfs/cbfs.go]: if you want, shorten this a little
if err blah; err != nil {

File: pkg/cbfs/fsp.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	"log"
+)
+
+func init() {
+	if err := RegisterFileReader(&SegReader{Type: TypeFSP, Name: "fsp", New: NewFSP}); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func NewFSP(f *File) (ReadWriter, error) {
```
Comment (rminnich) [inline pkg/cbfs/fsp.go:15]: you'll need to add comments for all this, but golint will doubtless complain for me :-)

File: pkg/cbfs/image.go [hunk 5]
```diff
@@ -116,31 +116,48 @@ func (i *Image) WriteFile(name string, perm os.FileMode) error {
...
 // by the fact that endianness is not consistent in cbfs images.
 func (i *Image) Update() error {
+	//FIXME: Support additional regions
 	// Because there can be gaps due to alignment of various
 	// components, we start out by filling i.Data with ff
 	// past the FMAP header.
-	for x := range i.Data[512:] {
-		i.Data[x+512] = 0xff
-	}
+	//for x := range i.Data[512:] {
```
Comment (rminnich) [inline pkg/cbfs/image.go]: remove this or uncomment it?

**Review by rminnich — COMMENTED** [review]
This is great. You'll need to run golint and go vet but it's a great start.

**Comment by hugelgupf** [comment]
It would be nice to have some tests for all these features as well

File: cmds/cbfs/cbfs.go (hunk 2 above)
Comment (PatrickRudolph) [reply cmds/cbfs/cbfs.go]: removed from this PR, feature will be added in the next PR as well as tests.

File: cmds/cbfs/cbfs.go (hunk 3 above)
Comment (PatrickRudolph) [reply cmds/cbfs/cbfs.go]: removed from this PR, feature will be added in the next PR as well as tests.

File: cmds/cbfs/cbfs.go (hunk 1 above)
Comment (PatrickRudolph) [reply cmds/cbfs/cbfs.go]: removed from this PR, feature will be added in the next PR as well as tests.

File: pkg/cbfs/fsp.go (hunk 4 above)
Comment (PatrickRudolph) [reply pkg/cbfs/fsp.go:15]: done

File: pkg/cbfs/image.go (hunk 5 above)
Comment (PatrickRudolph) [reply pkg/cbfs/image.go]: done

**Comment by PatrickRudolph** [comment]
Pushed fixes only. The next PR will contain features and tests.

File: cmds/cbfs/cbfs.go [hunk 6]
```diff
@@ -24,7 +24,7 @@ func main() {
 
 	i, err := cbfs.Open(a[0])
 	if err != nil {
-		log.Fatal(err)
+		log.Fatal("Failed to open file %s. Error:", a[0], err)
```
Comment (rminnich) [inline cmds/cbfs/cbfs.go]: The err will include the file name. In general, it's almost always safe to log.Fatal(err)

File: pkg/cbfs/empty.go [hunk 7]
```diff
@@ -18,6 +18,7 @@ func NewEmptyRecord(f *File) (ReadWriter, error) {
 	r := &EmptyRecord{File: *f}
 	Debug("Got header %v", r.String())
 	r.Type = TypeDeleted2
+	r.Name = ""
```
Comment (rminnich) [inline pkg/cbfs/empty.go]: no need to do this r.Name, that's how it will be initialized. This is not C, thank goodness :-)

File: pkg/cbfs/fsp.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+func (r *FSPRecord) String() string {
+	return recString(r.File.Name, r.RecordStart, r.Type.String(), r.Size, "none")
+}
+
+func (r *FSPRecord) Write(w io.Writer) error {
+	return Write(w, r.FData)
+}
+
+//Header returns a pointer to the corresponding File
+func (r *FSPRecord) Header() *File {
```
Comment (rminnich) [inline pkg/cbfs/fsp.go]: hmm, could it be called File then?

File: pkg/cbfs/image.go [hunk 9]
```diff
@@ -133,14 +128,20 @@ func (i *Image) Update() error {
 		if err := s.Write(&b); err != nil {
 			return err
 		}
-		Debug("Copy %d bytes to i.Data[%d]", len(b.Bytes()), s.Header().RecordStart+512)
-		copy(i.Data[s.Header().RecordStart+512:], b.Bytes())
+		end := uint32(len(b.Bytes())) + s.Header().RecordStart
+		if end > i.Area.Size {
+			return fmt.Errorf("write outside of CBFS not supported %x %x", end, i.Area.Size)
```
Comment (rminnich) [inline pkg/cbfs/image.go]: it won't be clear to me what the numbers mean and hence I'll be confused by this message. 
Could you say something like
"Region [%#x, %#x] is outside [%#x,#x]" Do regions have names? From this error I won't be able to tell which region it is.

File: pkg/cbfs/image.go [hunk 10]
```diff
@@ -162,9 +163,13 @@ func (i *Image) Remove(n string) error {
...
 		return os.ErrExist
 	}
-	// You can not remove the master header or the boot block.
+	// You can not remove the master header
 	// Just remake the cbfs if you're doing that kind of surgery.
-	if found == 0 || found == len(i.Segs)-1 {
+	if found == 0 {
+		return os.ErrPermission
+	}
+	// You can not remove the bootblock on x86
```
Comment (rminnich) [inline pkg/cbfs/image.go]: how do we know we're on an x86?

File: pkg/cbfs/spd.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	"log"
+)
+
+func init() {
+	if err := RegisterFileReader(&SegReader{Type: TypeSPD, Name: "spd", New: NewSPD}); err != nil {
+		log.Fatal(err)
+	}
+}
+
+//NewSPD returns a ReadWriter interface for the CBFS type TypeSPD
```
Comment (rminnich) [inline pkg/cbfs/spd.go]: you can leave out the word interface on this kind of description.

**Comment by ChriMarMe** [comment]
I'm taking over :)

**Comment by ChriMarMe** [comment]
ping @rminnich

**Comment by ChriMarMe** [comment]
ping @rminnich again :)

**Review by rminnich — APPROVED** [review]

**Pull request merged**